pkg/report/linux.go: slice bounds out of range in extractGuiltyFile()

Hi!
During linux kernel fuzzing on Debian 12, I get the error "slice bounds out of range" on kernel version 4.19.

Error log:
```bash
panic: runtime error: slice bounds out of range [131292:211]

goroutine 1 [running]:
github.com/google/syzkaller/pkg/report.(*linux).extractGuiltyFile(...)
        /kernel/syzkaller/pkg/report/linux.go:701
github.com/google/syzkaller/pkg/report.(*linux).Symbolize(0xc000558f20, 0xc008a282a0)
        /kernel/syzkaller/pkg/report/linux.go:388 +0x1ce
github.com/google/syzkaller/pkg/report.(*Reporter).Symbolize(0xc0006fd680, 0xc008a282a0)
        /kernel/syzkaller/pkg/report/report.go:219 +0x44
main.(*Manager).saveCrash(0xc003c824e0, 0xc001862e10)
        /kernel/syzkaller/syz-manager/manager.go:863 +0x4c
main.(*Manager).vmLoop(0xc003c824e0)
        /kernel/syzkaller/syz-manager/manager.go:424 +0x7e8
main.RunManager(0xc000ed2000)
        /kernel/syzkaller/syz-manager/manager.go:276 +0x92c
main.main()
        /kernel/syzkaller/syz-manager/manager.go:161 +0x12a
```

The error keeps repeating constantly after several hours of running the fuzzer. 
For some reason, the borders of the report are being calculated incorrectly.
An example of byte array (in source code: rep.Report) on which the error was triggered:
```go
data := []byte{109, 115, 103, 99, 116, 108, 36, 73, 80, 67, 95, 83, 69, 84, 40, 114, 49, 49, 44, 32, 48, 120, 49, 44, 32, 38, 40, 48, 120, 55, 102, 48, 48, 48, 48, 48, 48, 48, 52, 99, 48, 41, 61, 123, 123, 48, 120, 50, 44, 32, 114, 49, 50, 44, 32, 114, 49, 51, 44, 32, 48, 120, 101, 101, 48, 48, 44, 32, 48, 120, 101, 101, 48, 48, 44, 32, 48, 120, 52, 48, 44, 32, 48, 120, 49, 48, 49, 125, 44, 32, 48, 120, 48, 44, 32, 48, 120, 48, 44, 32, 48, 120, 49, 44, 32, 48, 120, 51, 44, 32, 48, 120, 51, 44, 32, 48, 120, 49, 44, 32, 48, 120, 52, 53, 48, 101, 44, 32, 48, 120, 102, 102, 44, 32, 48, 120, 52, 48, 49, 44, 32, 48, 120, 51, 44, 32, 48, 120, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 102, 44, 32, 114, 49, 52, 125, 41, 10, 77, 67, 69, 58, 32, 75, 105, 108, 108, 105, 110, 103, 32, 109, 111, 100, 112, 114, 111, 98, 101, 58, 51, 48, 55, 57, 50, 32, 100, 117, 101, 32, 116, 111, 32, 104, 97, 114, 100, 119, 97, 114, 101, 32, 109, 101, 109, 111, 114, 121, 32, 99, 111, 114, 114, 117, 112, 116, 105, 111, 110, 32, 102, 97, 117, 108, 116, 32, 97, 116, 32, 55, 102, 101, 99, 100, 101, 49, 100, 52, 48, 48, 48, 10}
```
Fuzzer run command:
```bash
# inside syzkaller dir
./bin/syz-manager -config /kernel/4_19_261/my.cfg
```

Host OS:
```bash
root@workspace:~# uname -a
Linux workspace 6.1.0-10-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.38-2 (2023-07-27) x86_64 GNU/Linux
```
Config:
```json
{
        "target": "linux/amd64",
        "http": "0.0.0.0:80",
        "workdir": "/kernel/4_19_261/workdir",
        "kernel_obj": "/root/rpmbuild/BUILD/linux-4.19.261",
        "image": "/kernel/image/bullseye.img",
        "sshkey": "/kernel/image/bullseye.id_rsa",
        "syzkaller": "/kernel/syzkaller",
        "reproduce": false,
        "procs": 24,
        "type": "qemu",
        "vm": {
                "count": 8,
                "kernel": "/root/rpmbuild/BUILD/linux-4.19.261/arch/x86/boot/bzImage",
                "cpu": 2,
                "mem": 2048
        }
}
```

bullseye.img generated via [create-image.sh](https://github.com/google/syzkaller/blob/master/tools/create-image.sh).

The newly added assertions have fired on my local syzkaller instance:

```
panic: invalid reportPrefixLen after symbolize: prefix 186651 -> 186651, report len: 186776 -> 274, report: crash: SYZFATAL: executor NUM failed NUM times: executor NUM: EOF
```

So this code is making the report much much shorter

https://github.com/google/syzkaller/blob/c8349e48534ea6d8f01515335d95de8ebf5da8df/pkg/report/linux.go#L414-L426

--- a/pkg/report/bsd.go
+++ b/pkg/report/bsd.go
@@ -4,7 +4,6 @@
 package report
 
 import (
-	"bufio"
 	"bytes"
 	"fmt"
 	"path/filepath"
@@ -57,11 +56,9 @@ func (ctx *bsd) Symbolize(rep *Report) error {
 	symb := symbolizer.NewSymbolizer(ctx.config.target)
 	defer symb.Close()
 	var symbolized []byte
-	s := bufio.NewScanner(bytes.NewReader(rep.Report))
 	prefix := rep.reportPrefixLen
-	for s.Scan() {
-		line := append([]byte{}, s.Bytes()...)
-		line = append(line, '\n')
+	for _, line := range bytes.SplitAfter(rep.Report, []byte("\n")) {
+		line := bytes.Clone(line)
 		newLine := ctx.symbolizeLine(symb.Symbolize, line)
 		if prefix > len(symbolized) {
 			prefix += len(newLine) - len(line)

--- a/pkg/report/fuchsia.go
+++ b/pkg/report/fuchsia.go
@@ -4,7 +4,6 @@
 package report
 
 import (
-	"bufio"
 	"bytes"
 	"fmt"
 	"path/filepath"
@@ -97,8 +96,7 @@ func (ctx *fuchsia) shortenStarnixPanicReport(report []byte, maxUnrelatedLines, 
 	out := new(bytes.Buffer)
 	unrelatedLines := 0
 	unmatchedLines := 0
-	for s := bufio.NewScanner(bytes.NewReader(report)); s.Scan(); {
-		line := s.Bytes()
+	for _, line := range lines(report) {
 		if matchesAny(line, starnixFramePatterns) {
 			unrelatedLines = 0
 			unmatchedLines = 0
@@ -113,16 +111,16 @@ func (ctx *fuchsia) shortenStarnixPanicReport(report []byte, maxUnrelatedLines, 
 		out.Write(line)
 		out.WriteByte('\n')
 		if unmatchedLines == maxUnmatchedLines {
-			return out.Bytes()
+			break
 		}
 	}
-	return out.Bytes()
+	return append(bytes.TrimRight(out.Bytes(), "\n"), '\n')
 }
 
 func (ctx *fuchsia) shortenReport(report []byte) []byte {
 	out := new(bytes.Buffer)
-	for s := bufio.NewScanner(bytes.NewReader(report)); s.Scan(); {
-		line := zirconLinePrefix.ReplaceAll(s.Bytes(), nil)
+	for _, rawLine := range lines(report) {
+		line := zirconLinePrefix.ReplaceAll(rawLine, nil)
 		if matchesAny(line, zirconUnrelated) {
 			continue
 		}
@@ -139,14 +137,17 @@ func (ctx *fuchsia) symbolize(output []byte) []byte {
 	symb := symbolizer.NewSymbolizer(ctx.config.target)
 	defer symb.Close()
 	out := new(bytes.Buffer)
-	for s := bufio.NewScanner(bytes.NewReader(output)); s.Scan(); {
-		line := s.Bytes()
+
+	lines := lines(output)
+	for i := 0; i < len(lines); i++ {
+		line := lines[i]
 		if bytes.Contains(line, zirconAssertFailed) && len(line) == 127 {
 			// This is super hacky: but zircon splits the most important information in long assert lines
 			// (and they are always long) into several lines in irreversible way. Try to restore full line.
 			line = append([]byte{}, line...)
-			if s.Scan() {
-				line = append(line, s.Bytes()...)
+			if i+1 < len(lines) {
+				line = append(bytes.Clone(line), lines[i+1]...)
+				i++
 			}
 		}
 		if ctx.obj != "" {

--- a/pkg/report/linux.go
+++ b/pkg/report/linux.go
@@ -375,15 +375,7 @@ func (ctx *linux) Symbolize(rep *Report) error {
 			return err
 		}
 	}
-
-	oldLen := len(rep.Report)
 	rep.Report = ctx.decompileOpcodes(rep.Report, rep)
-	if len(rep.Report) > 0 && rep.reportPrefixLen > len(rep.Report) {
-		// An attempt to catch #4198.
-		panic(fmt.Sprintf("invalid reportPrefixLen (%d) after decompileOpcodes, report len: %d -> %d, report: %+v",
-			rep.reportPrefixLen, oldLen, len(rep.Report), rep,
-		))
-	}
 
 	// Skip getting maintainers for Android fuzzing since the kernel source
 	// directory structure is different.
@@ -411,27 +403,24 @@ func (ctx *linux) symbolize(rep *Report) error {
 		return ctx.symbolizerCache.Symbolize(symb.Symbolize, bin, pc)
 	}
 	var symbolized []byte
-	s := bufio.NewScanner(bytes.NewReader(rep.Report))
 	prefix := rep.reportPrefixLen
-	for s.Scan() {
-		line := append([]byte{}, s.Bytes()...)
-		line = append(line, '\n')
+	for _, originalLine := range bytes.SplitAfter(rep.Report, []byte("\n")) {
+		line := append([]byte{}, originalLine...)
 		newLine := symbolizeLine(symbFunc, ctx.symbols, ctx.vmlinux, ctx.kernelBuildSrc, line)
 		if prefix > len(symbolized) {
 			prefix += len(newLine) - len(line)
 		}
 		symbolized = append(symbolized, newLine...)
 	}
-	oldLen := len(rep.Report)
+	oldReport := rep.Report
 	rep.Report = symbolized
 	oldPrefixLen := rep.reportPrefixLen
 	rep.reportPrefixLen = prefix
 
 	if len(rep.Report) > 0 && rep.reportPrefixLen > len(rep.Report) {
-		// An attempt to catch #4198.
 		panic(fmt.Sprintf("invalid reportPrefixLen after symbolize: prefix %d -> %d,"+
-			"report len: %d -> %d, report: %+v",
-			oldPrefixLen, rep.reportPrefixLen, oldLen, len(rep.Report), rep,
+			"report len: %d -> %d, old report: %q",
+			oldPrefixLen, rep.reportPrefixLen, len(oldReport), len(rep.Report), oldReport,
 		))
 	}
 	return nil

--- a/pkg/report/linux.go
+++ b/pkg/report/linux.go
@@ -4,7 +4,6 @@
 package report
 
 import (
-	"bufio"
 	"bytes"
 	"encoding/binary"
 	"fmt"
@@ -404,8 +403,8 @@ func (ctx *linux) symbolize(rep *Report) error {
 	}
 	var symbolized []byte
 	prefix := rep.reportPrefixLen
-	for _, originalLine := range bytes.SplitAfter(rep.Report, []byte("\n")) {
-		line := append([]byte{}, originalLine...)
+	for _, line := range bytes.SplitAfter(rep.Report, []byte("\n")) {
+		line := bytes.Clone(line)
 		newLine := symbolizeLine(symbFunc, ctx.symbols, ctx.vmlinux, ctx.kernelBuildSrc, line)
 		if prefix > len(symbolized) {
 			prefix += len(newLine) - len(line)
@@ -658,8 +657,12 @@ func (ctx *linux) decompileOpcodes(text []byte, report *Report) []byte {
 	// Iterate over all "Code: " lines and pick the first that could be decompiled
 	// that might be of interest to the user.
 	var decompiled *decompiledOpcodes
-	var prevLine []byte
-	for s := bufio.NewScanner(bytes.NewReader(text)); s.Scan(); prevLine = append([]byte{}, s.Bytes()...) {
+	lines := lines(text)
+	for i, line := range lines {
+		var prevLine []byte
+		if i > 0 {
+			prevLine = lines[i-1]
+		}
 		// We want to avoid decompiling code from user-space as it is not of big interest during
 		// debugging kernel problems.
 		// For now this check only works for x86/amd64, but Linux on other architectures supported
@@ -667,7 +670,7 @@ func (ctx *linux) decompileOpcodes(text []byte, report *Report) []byte {
 		if linuxUserSegmentRe.Match(prevLine) {
 			continue
 		}
-		match := linuxCodeRe.FindSubmatch(s.Bytes())
+		match := linuxCodeRe.FindSubmatch(line)
 		if match == nil {
 			continue
 		}
@@ -731,12 +734,13 @@ func (ctx *linux) extractGuiltyFileRaw(title string, report []byte) string {
 }
 
 func (ctx *linux) extractGuiltyFileImpl(report []byte) string {
-	scanner := bufio.NewScanner(bytes.NewReader(report))
-
 	// Extract the first possible guilty file.
 	guilty := ""
-	for scanner.Scan() {
-		match := filenameRe.FindSubmatch(scanner.Bytes())
+	var line []byte
+	lines := lines(report)
+	for len(lines) > 0 {
+		line, lines = lines[0], lines[1:]
+		match := filenameRe.FindSubmatch(line)
 		if match == nil {
 			continue
 		}
@@ -748,7 +752,7 @@ func (ctx *linux) extractGuiltyFileImpl(report []byte) string {
 			guilty = string(file)
 		}
 
-		if matchesAny(file, ctx.guiltyFileIgnores) || ctx.guiltyLineIgnore.Match(scanner.Bytes()) {
+		if matchesAny(file, ctx.guiltyFileIgnores) || ctx.guiltyLineIgnore.Match(line) {
 			continue
 		}
 		guilty = filepath.Clean(string(file))
@@ -757,13 +761,14 @@ func (ctx *linux) extractGuiltyFileImpl(report []byte) string {
 
 	// Search for deeper filepaths in the stack trace below the first possible guilty file.
 	deepestPath := filepath.Dir(guilty)
-	for scanner.Scan() {
-		match := filenameRe.FindSubmatch(scanner.Bytes())
+	for len(lines) > 0 {
+		line, lines = lines[0], lines[1:]
+		match := filenameRe.FindSubmatch(line)
 		if match == nil {
 			continue
 		}
 		file := match[1]
-		if matchesAny(file, ctx.guiltyFileIgnores) || ctx.guiltyLineIgnore.Match(scanner.Bytes()) {
+		if matchesAny(file, ctx.guiltyFileIgnores) || ctx.guiltyLineIgnore.Match(line) {
 			continue
 		}
 		clean := filepath.Clean(string(file))
@@ -839,7 +844,7 @@ func (ctx *linux) isCorrupted(title string, report []byte, format oopsFormat) (b
 		if match == nil {
 			continue
 		}
-		frames := bytes.Split(report[match[0]:], []byte{'\n'})
+		frames := lines(report[match[0]:])
 		if len(frames) < 4 {
 			return true, "call trace is missed"
 		}

--- a/pkg/report/report.go
+++ b/pkg/report/report.go
@@ -6,7 +6,6 @@
 package report
 
 import (
-	"bufio"
 	"bytes"
 	"fmt"
 	"regexp"
@@ -585,9 +584,13 @@ func extractStackFrame(params *stackParams, stack *stackFmt, output []byte) ([]s
 	return extractStackFrameImpl(params, output, skipRe, stack.parts2, extractor)
 }
 
+func lines(text []byte) [][]byte {
+	return bytes.Split(text, []byte("\n"))
+}
+
 func extractStackFrameImpl(params *stackParams, output []byte, skipRe *regexp.Regexp,
 	parts []*regexp.Regexp, extractor frameExtractor) ([]string, bool) {
-	s := bufio.NewScanner(bytes.NewReader(output))
+	lines := lines(output)
 	var frames, results []string
 	ok := true
 	numStackTraces := 0
@@ -607,8 +610,9 @@ nextPart:
 		part := parts[partIdx]
 		if part == parseStackTrace {
 			numStackTraces++
-			for s.Scan() {
-				ln := s.Bytes()
+			var ln []byte
+			for len(lines) > 0 {
+				ln, lines = lines[0], lines[1:]
 				if matchesAny(ln, params.corruptedLines) {
 					ok = false
 					continue nextPart
@@ -635,8 +639,9 @@ nextPart:
 				frames = appendStackFrame(frames, match, params, skipRe)
 			}
 		} else {
-			for s.Scan() {
-				ln := s.Bytes()
+			var ln []byte
+			for len(lines) > 0 {
+				ln, lines = lines[0], lines[1:]
 				if matchesAny(ln, params.corruptedLines) {
 					ok = false
 					continue nextPart

